qutip/settings.py

```diff
@@ -84,3 +85,267 @@
     del logging  # Don't leak names!
 except:
     _logger = None
+
+
+def _valid_config(key):
+    if key == "absolute_import":
+        return False
+    if key.startswith("_"):
+        return False
+    val = _self[key]
+    if isinstance(val, (bool, int, float, complex, str)):
+        return True
+    return False
+
+
+_environment_keys = ["ipython", 'has_mkl', 'has_openmp',
+                     'mkl_lib', 'fortran', 'num_cpus']
+_self = locals().copy()  # Not ideal, making an object would be better
```

Comment:
I agree an object would be better.

The `Options` classes could support a proper hierarchy, which might be very nice. Then this module can define the "root" object, and everything else can just attach to it.  This could also solve the circular import problem with `configrc` - you can remove the section specification from `configrc`, make that module completely private, and have all of the functions take or return one of these tree objects.  The user only accesses `save` and `load` via `qutip.settings` anyway.

We could do the hierarchy without worrying about duplicating names by effectively splitting each class into two scopes, so that "dictionary" access accesses individual settings, and "dot" access walks the hierarchy.  So changing settings becomes `settings["colorblind_safe"] = False`, `settings.solver["atol"] = 1e-12`, or even `settings.solver.mcsolve["ntraj"] = 500`?

Reply:
I would like I am not sure I fully follow you here. But if you are saying the individual setting are no longer in the module `qutip.settings` but in an object `setting`. That this object is the only one with knowledge of `configrc` (and not the other way around). Yes that would be nice for v5. For now, I allow the instance of options to have access to `save` and `load`. Since user are used to pass `Options` object around. Saving and reusing it will come more naturally than changing the default.

For the dictionary access, i am so use of attribute access that they seems weird, but having a clear distinction between the flag and the hierarchy seems logic now that we have one. I would like user feedback on this.

Reply:
Er yeah, I guess that was a bit confusing because I called the object and the module the same thing.

The main thing I meant to say was about how there _could_ be a distinction between hierarchy access and settings access.  I'm not sure which way I'd even want to go.  Having two different "namespaces" (for hierarchy and for settings) makes it easier to avoid naming clashes, but unifying them looks rather cleaner to the user.  I suppose that you can assign an entire `Options` class to one part of the hierarchy (e.g. `settings.solver = SolverOptions(...)`) anyway, so having the separate dictionary syntax would be a weird split if you did that.  Asking over people is definitely the way to go.